### PR TITLE
[Merged by Bors] - feat(Algebra/Ring/Ext): prove extensionality lemmas for Ring and similar typeclasses

### PR DESCRIPTION
**Pull request by raghuram-13**
Add a file `Algebra/Ring/Ext`, proving `ext` lemmas for all of the ring-like classes (for example, anything from `NonUnitalNonAssocSemiring` to `CommRing`), stating that two instances of such a class are equal if they define the same addition and multiplication operations.
Also prove `ext_iff` variants for each class, and injectivity lemmas for projections from classes to parent classes.

---
Some notes:
+ Things I didn't do because they don't seem particularly necessary:
  + I skipped `HasDistribNeg`.
  + More injectivity lemmas could be proved between the `Comm` variants (currently it's just from `Comm` variant to non-`Comm` variant for each).
+ A natural next step would be to extend this to `DivisionSemiring`, `DivisionRing`, `Semifield` and `Field`.
+ I have introduced local notation to eliminate some boilerplate. I'm not sure if this is style-appropriate for Mathlib (although the notation is local to one file and (probably) never shows up when e.g. the types of lemmas are viewed, so that it's invisible to anyone not editing this file).
+ In [another branch](https://github.com/leanprover-community/mathlib4/raghuram-13/ext-for-rings.experimental), I have used a (local) custom command to generate the `ext` and `ext_iff` lemmas, eliminating even more boilerplate.

  It should be possible to extend that command to be more general (such as allowing `declModifiers` and `terminationSuffix` clauses, allowing a custom theorem name defaulting to `ext`, specifying the type for which an ext lemma is being proved, specifying the arguments, and letting the names be inaccessible by default while allowing them to be specified with a "with" clause), in which case it could become a useful command for writing extensionality lemmas (admittedly partially overlapping in functionality with putting an `@[ext]` on the type definition).

+ I haven't written any tests — I'm not sure what to test or how to write tests.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Ext.lean:44]: In the other ext lemmas, we write
```suggestion
    (h_mul : (letI := inst₁; HAdd.hAdd : R → R → R) = (letI := inst₂; HAdd.hAdd : R → R → R)) 
```
Can we do the same here?

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Ext.lean:57]: We don't have these `ext_iff` versions for `Monoid` and friends (and I suspect they're not actually useful), so I'd be inclined to skip them for now.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Ext.lean:114]: We should certainly add an ext lemma for `AddMonoidWithOne`

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Ext.lean:21]:
```suggestion
multiplication specified by a typeclass instance `i` on a type `R` (and similarly for addition). We
```

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Ring/Ext.lean:24]:
```suggestion
Since `Mathlib/Algebra/Group/Ext` proved several injectivity lemmas, we do so as well — even if sometimes we
```

Comment (riccardobrasca) [reply Mathlib/Algebra/Ring/Ext.lean:24]: Maybe with `.` instead of `/`?

Comment (eric-wieser) [reply Mathlib/Algebra/Ring/Ext.lean:24]: I think `Mathlib/Algebra/Group/Ext.lean` is probably the best bet, since that is easier to detect heuristically in doc-gen for linkification.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:44]: I thought that in practice equality of functions would mostly be proved by extensionality and so this would "shift" the need to use `funext` from uses to here, making the lemmas more usable.

For reference, I found two uses of the extensionality lemmas from `Mathlib.Algebra.Group.Ext`, in both of which the argument is proved using `funext`:
- `Group.ext` used in `Mathlib.Topology.Homotopy.HomotopyGroup`, around line 526
-  `AddCommGroup.ext` used in `Mathlib.CategoryTheory.Preadditive.Biproducts`, around line 657

Of course, I can also go ahead and change everything to "function equality" from "application equality".

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:114]: Sure; where though?
After the definition of `AddMonoidWithOne` in `Mathlib/Data/Nat/Cast/Defs.lean`, or in a new `Ext.lean` file in the same folder, or in `Mathlib/Algebra/Group/Ext.lean`, or in this file `Mathlib/Algebra/Ring/Ext.lean`? (Or anywhere else?)

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:232]: Same question as for `AddMonoidWithOne`.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:24]: IIRC I copied this style of referring to files from the commit conventions.

OTOH, I see the `.` style in the module header for `Algebra.Group.Defs` (without the `Mathlib` prefix), the `/` style (without `Mathlib` prefix, with `.lean` suffix) in `Algebra/Group/Basic.lean` and `Algebra/Group/UniqueProds.lean`, etc.

It's also noteworthy that the [contribution guidelines for documentation](https://leanprover-community.github.io/contribute/doc.html) mention that fully-qualified names (e.g., `Group.ext`) are turned into links to the online docs, so that could cause a conflict if `.`-style notation is used to refer to modules and there is a definition and a module with the same name eventually (although that's unlikely if the `Mathlib` prefix is used for the module names).

So which style should I go with?

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:44]: By the way, I noticed that `Group.ext` and `CommGroup.ext` in `Mathlib/Algebra/Group/Ext.lean` use `g.mul` rather than `(letI := g; HMul.hMul: G → G → G)`, which goes against what is written at the top of that file.

Should this be changed?

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:232]: And here's an implementation to use later as with `AddMonoidWithOne`:
```lean
@[ext] theorem AddGroupWithOne.ext ⦃inst₁ inst₂ : AddGroupWithOne R⦄
    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
    (h_one : (letI := inst₁; One.one : R) = (letI := inst₂; One.one)) :
    inst₁ = inst₂ := by
  have : inst₁.toAddMonoidWithOne = inst₂.toAddMonoidWithOne :=
    AddMonoidWithOne.ext h_add h_one
  have : inst₁.toNatCast = inst₂.toNatCast := congrArg (·.toNatCast) this
  have h_group : inst₁.toAddGroup = inst₂.toAddGroup := by ext; apply h_add
  -- Extract equality of necessary substructures from h_group
  injection h_group with h_group; injection h_group
  have : inst₁.toIntCast.intCast = inst₂.toIntCast.intCast := by
    funext n; cases n with
    | ofNat n   => rewrite [← Int.coe_nat_eq, inst₁.intCast_ofNat, inst₂.intCast_ofNat]; congr
    | negSucc n => rewrite [inst₁.intCast_negSucc, inst₂.intCast_negSucc]; congr
  rcases inst₁ with @⟨⟨⟩⟩; rcases inst₂ with @⟨⟨⟩⟩
  congr

@[ext] theorem AddCommGroupWithOne.ext ⦃inst₁ inst₂ : AddCommGroupWithOne R⦄
    (h_add : ∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
    (h_one : (letI := inst₁; One.one : R) = (letI := inst₂; One.one)) :
    inst₁ = inst₂ := by
  have : inst₁.toAddCommGroup = inst₂.toAddCommGroup :=
    AddCommGroup.ext (funext₂ h_add)
  have : inst₁.toAddGroupWithOne = inst₂.toAddGroupWithOne :=
    AddGroupWithOne.ext h_add h_one
  injection this with _ h_addMonoidWithOne; injection h_addMonoidWithOne
  cases inst₁; cases inst₂
  congr
```

(And the change in `NonAssocRing.ext`.)
```diff
   have h₂ : inst₁.toNonAssocSemiring = inst₂.toNonAssocSemiring := by
     ext <;> apply_assumption
-  -- Mathematically non-trivial fact: `intCast` is determined by the rest.
-  have h_intCast : inst₁.toIntCast.intCast = inst₂.toIntCast.intCast := by
-    have : inst₁.toNatCast = inst₂.toNatCast := by injection h₂
-    funext n; cases n with
-    | ofNat n   => rewrite [← Int.coe_nat_eq, inst₁.intCast_ofNat, inst₂.intCast_ofNat]; congr
-    | negSucc n => rewrite [inst₁.intCast_negSucc, inst₂.intCast_negSucc]; congr
-  -- Split into fields (extracting `intCast` function) and prove they are equal using the above.
-  rcases inst₁ with @⟨_, _, _, _, _, _, _, ⟨⟩⟩
-  rcases inst₂ with @⟨_, _, _, _, _, _, _, ⟨⟩⟩
-  congr <;> try solve| injection h₁ | injection h₂
+  have h₃ : inst₁.toAddCommGroupWithOne = inst₂.toAddCommGroupWithOne :=
+    AddCommGroupWithOne.ext h_add (congrArg (·.toOne.one) h₂)
+  cases inst₁; cases inst₂
+  congr <;> solve| injection h₁ | injection h₂ | injection h₃
```

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:57]: I added them because the `ext` attribute generates them for `class` definitions (probably `structure` and `inductive` as well).

Searching through Mathlib does turn up several uses of `ext_iff` lemmas, both as `ext_iff.1`/`ext_iff.mp` (which I guess is essentially a convenient way to avoid having to incant the right `congr` or `injection`s) and in `simp`s, although I don't know if these ones are going to be of any use.

I guess there's no harm in keeping these, even if we don't keep writing them?

One thing I did notice is that for `ext_iff` to be usable as `ext_iff.1`/`ext_iff.mp`, it should only take instance or implicit arguments (although this is not the behaviour of `@[ext]`-generated `ext_iff` lemmas).
If these `ext_iff` are kept, they could be changed to do the same (or not).

**Comment by j-loreaux** [comment]
@raghuram-13 just FYI: maintainers generally use the "assignees" feature on GitHub to communicate to other maintainers that they have taken responsibility for deciding when to merge the PR. Therefore, if you add yourself as an assignee, it might have the unfortunate side effect of preventing your PRs from getting reviewed.

**Comment by raghuram-13** [comment]
> @raghuram-13 just FYI: maintainers generally use the "assignees" feature on GitHub to communicate to other maintainers that they have taken responsibility for deciding when to merge the PR. Therefore, if you add yourself as an assignee, it might have the unfortunate side effect of preventing your PRs from getting reviewed.

I thought that meant I was working on it :facepalm:. Oops.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (urkud) [reply Mathlib/Algebra/Ring/Ext.lean:44]: What lemma does `@[ext]` generate for `Distrib`? Is it the same?

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [reply Mathlib/Algebra/Ring/Ext.lean:114]: > or in this file Mathlib/Algebra/Ring/Ext.lean?

I think this is fine. Anywhere is better than in a PR comment!

File: Mathlib/Algebra/Ring/Ext.lean
Comment (eric-wieser) [reply Mathlib/Algebra/Ring/Ext.lean:24]: Let's use the `/` style for filenames. It's very easy to change that to dot notation later in a mathlib-wide sweep, but changing in the other direction is much harder.

**Comment by eric-wieser** [comment]
Sorry for dropping this; your diffs in the PR comments above generally look good; please apply them!

**Comment by raghuram-13** [comment]
Sorry for the hiatus; I'll implement the changes now.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:44]: The `@[ext]` in my code just registers the `ext` lemma I have proved for the `ext` tactic, as far as I know.

In case you mean what an `@[ext]` added to the definition of `Distrib` would do, it appears to be something more along the lines of `inst₁.toMul = inst₂.toMul → inst₁.toAdd = inst₂.toAdd → inst₁ = inst₂`.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:143]: Done in 309b6ed6da.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean]: Added in 309b6ed6da.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean]: Done in 10abca9012.

**Comment by mattrobball** [comment]
@eric-wieser could please approve if appropriate? Thanks.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [inline Mathlib/Algebra/Ring/Ext.lean:8]: A commented-out line that is no longer needed can be safely deleted (we can always check the commits to get it back).
```suggestion
```

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [reply Mathlib/Algebra/Ring/Ext.lean:44]: I agree with @urkud that we should follow the format of the `Distrib.ext` lemma which is generated by `@[ext] class Distrib`. In fact, [there is a library `note [partially-applied ext lemmas]` about this](https://leanprover-community.github.io/mathlib_docs/notes.html#partially-applied%20ext%20lemmas) (sadly never ported to Lean 4):

> When possible, `ext` lemmas are stated without a full set of arguments. As an example, for bundled
homs `f`, `g`, and `of`, `f.comp of = g.comp of → f = g` is a better `ext` lemma than
`(∀ x, f (of x) = g (of x)) → f = g`, as the former allows a second type-specific extensionality
lemmas to be applied to `f.comp of = g.comp of`.
If the domain of `of` is `ℕ` or `ℤ` and `of` is a `ring_hom`, such a lemma could then make the goal
`f (of 1) = g (of 1)`.
> 
> For bundled morphisms, there is a `ext` lemma that always applies of the form
`(∀ x, ⇑f x = ⇑g x) → f = g`. When adding type-specific `ext` lemmas like the one above, we want
these to be tried first. This happens automatically since the type-specific lemmas are inevitably
defined later.

So in this case we would use the unapplied `HAdd.add` and `HMul.mul`. (Note that `@[ext] class Distrib` is not quite right: it uses `Add.add` and `Mul.mul`, which means we don't get the nice notation.)

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [reply Mathlib/Algebra/Ring/Ext.lean:57]: Now that they are written, I don't see much harm in keeping the `ext_iff`s. In this case the autogeneration behaviour of `@[ext]` is probably the wrong thing to do: as a rule of thumb, arguments that appear on both sides of an `iff` can be safely left implicit.

Here's another way to write the proof. Using `rintro rfl` means both sides are immediately equal, so no `congr` is needed. But either way is fine really.
```suggestion
theorem ext_iff (inst₁ inst₂ : Distrib R) :
    inst₁ = inst₂ ↔
      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y)
      ∧ (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
  ⟨by rintro rfl; constructor <;> (intros; rfl), And.elim (ext · ·)⟩
```

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [inline Mathlib/Algebra/Ring/Ext.lean:56]: Nitpick: we put operators before a linebreak. Could you also make the same change below?
```suggestion
      (∀ x y, x +[R, inst₁] y = x +[R, inst₂] y) ∧
        (∀ x y, x *[R, inst₁] y = x *[R, inst₂] y) :=
```

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [inline Mathlib/Algebra/Ring/Ext.lean:112]: Would it work to move these results to `Mathlib/Algebra/Group/WithOne/Basic.lean`? The same for `AddGroupWithOne`.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [inline Mathlib/Algebra/Ring/Ext.lean:110]: Please move this comment to `section NonAssocSemiring`.

**Review by Vierkantor — COMMENTED** [review]
Sorry for the long waiting time before the new review! I assume Eric had been very busy with finishing his thesis in the last weeks. Unfortunately there is still some work to be done with this pull request.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [reply Mathlib/Algebra/Ring/Ext.lean:44]: In fact Eric has an in-depth explanation of why we shouldn't apply `ext` lemmas in his dissertation: https://eric-wieser.github.io/thesis/eric-wieser-thesis.pdf (chapter 5).

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:8]: Must have forgotten to remove it long ago!

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean]: Done in 6580ed52468b94fd5c04718c6bde7b45564511d5.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:109]: It's currently here because until the three lemmas following are relocated, they are functionally auxiliary lemmas for `NonAssocSemiring` extensionality. When those are moved, this will end up being at `namespace NonAssocSemiring` as desired.

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean]: That's a very good point for bundled morphisms, but I'm not sure that it applies to functions, where passing back from an "applied" goal `forall x y, f x y = g x y` to an "unapplied" goal `f = g` is a couple of `congr`s or so away for automation (unlike a bundled morphism, where one might have to reconstruct the bundled stuff if it has been simplified away), and as I mentioned above, it does seem like the "applied" goal is more often desired when used manually.

But in any case, I have changed everything to use "unapplied" ext subgoals in fbee3e7 (it was just a find-and-replace followed by patching proofs, but I don't *think* there are any errors in it now; still waiting for CI to finish).
(If this ever needs to be changed back, that and the three commits before it can be reverted.)

File: Mathlib/Algebra/Ring/Ext.lean
Comment (raghuram-13) [reply Mathlib/Algebra/Ring/Ext.lean:111]: `Mathlib/Algebra/Group/WithOne` appears to be about an operator `WithOne` which adjoins a new zero/one to a type/algebraic structure, rather than about algebraic structures which already have one.

The `AddMonoidWithOne` etc. about which we have extra ext lemmas here are all defined in `Mathlib/Data/Nat/Cast/Defs.lean` or `Mathlib/Data/Int/Cast/Defs.lean`. As to where that means their ext lemmas should be placed, :shrug:. There is a resemblance in name to the classes of `Mathlib/Algebra/GroupWithZero`...

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [reply Mathlib/Algebra/Ring/Ext.lean:109]: I understand your point. However, pull requests should be self-contained, so comments etc. should reflect the actual state of the code, and not the proposed future state.

How about:
```suggestion
/-! ### NonAssocSemiring and its ancestors

This section also includes results for `AddMonoidWithOne`, `AddCommMonoidWithOne`, etc.
as these are considered implementation detail of the ring classes.
TODO consider relocating these lemmas.
-/
```

File: Mathlib/Algebra/Ring/Ext.lean
Comment (Vierkantor) [inline Mathlib/Algebra/Ring/Ext.lean:240]:
```suggestion
/-! ### NonAssocRing and its ancestors

This section also includes results for `AddGroupWithOne`, `AddCommGroupWithOne`, etc.
as these are considered implementation detail of the ring classes.
TODO consider relocating these lemmas. -/
```

**Review by Vierkantor — APPROVED** [review]
Sorry for the late response, this dropped somewhat off my radar. The code looks good to me. Still have some small remarks about the comments.

bors d+

**Pull request closed without merging**
